### PR TITLE
Event buffer for structlog

### LINKED ISSUE #4320: Scalable in-memory event history
Does something graceful when it runs out of memory (small machines shouldn't explode because of logging), is queryable by a potential future server-wrapper.

### PR DESCRIPTION
**Pull request by iknox-fa**
resolves #4320

### Description
Adds a global event history buffer using a FIFO strategy to keep memory usage reasonable.  Can be used when wrapping dbt or in general to access past events.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [] I have updated the `CHANGELOG.md` and added information about my change

### CODE REVIEW

File: core/dbt/events/functions.py [hunk 1]
```diff
@@ -16,9 +16,14 @@
...
 import uuid
-from typing import Any, Callable, Dict, List, Optional, Set, Tuple, Union
+from typing import Any, Callable, Dict, List, Optional, Set, Tuple, Union, Deque
 from dataclasses import _FIELD_BASE  # type: ignore[attr-defined]
+from collections import deque
 
+# create the global event history buffer with a max size of 100k records
+# TODO: make the maxlen something configurable from the command line via args(?)
+global EVENT_HISTORY
+EVENT_HISTORY: Deque = deque(maxlen=100000)
```
Comment (iknox-fa) [inline core/dbt/events/functions.py:26]: @kwigley is this going to fit the bill for what's needed in Client/Server?  It should make a globally available deque of events so you can do whatever's needed.

File: core/dbt/events/functions.py [hunk 2]
```diff
@@ -281,7 +286,15 @@ def send_exc_to_logger(
...
 # to files, etc.)
 def fire_event(e: Event) -> None:
-    # TODO manage history in phase 2:  EVENT_HISTORY.append(e)
+
+    # if and only if the event history deque will be completely filled by this event
+    # fire warning that old events are now being dropped
+    global EVENT_HISTORY
+    if len(EVENT_HISTORY) == (100000 - 1):
+        EVENT_HISTORY.append(e)
+        fire_event(EventBufferFull())
```
Comment (nathaniel-may) [inline core/dbt/events/functions.py:295]: what if someone is running on a machine so small that memory runs out before this hard-coded limit? would we never log it?

File: core/dbt/events/functions.py [hunk 3]
```diff
@@ -281,7 +286,15 @@ def send_exc_to_logger(
...
+    else:
```
Comment (nathaniel-may) [inline core/dbt/events/functions.py:296]: I'd prefer this if/else to be in the form:
```python
if we_need_to_log:
    log()

append()
```

File: test/unit/test_events.py [hunk 4]
```diff
@@ -70,4 +74,24 @@ def test_event_codes(self):
...
+    def test_buffer_populates(self):
+        fire_event(UnitTestInfo(msg="Test Event 1"))
+        fire_event(UnitTestInfo(msg="Test Event 2"))
+        self.assertTrue(
+            EVENT_HISTORY.count(UnitTestInfo(msg='Test Event 1', code='T006')) == 1
+        )
+
+    # ensure events drop from the front of the buffer when buffer maxsize is reached
+    def test_buffer_FIFOs(self):
+        for n in range(0,100001):
```
Comment (iknox-fa) [inline test/unit/test_events.py:90]: FYI -- This is adding ~10 seconds to the unit test run time... can be fixed by correcting TODO on `core/dbt/events/functions.py` line 24.

**Review by nathaniel-may — CHANGES_REQUESTED** [review]
one question and one minor change

File: test/unit/test_events.py (hunk 4 above)
Comment (nathaniel-may) [reply test/unit/test_events.py:90]: ahh. making it configurable so we can test with a smaller queue.

File: core/dbt/events/functions.py (hunk 1 above)
Comment (kwigley) [reply core/dbt/events/functions.py:26]: awesome! Ya, I believe so.

**Review by leahwicz — APPROVED** [review]

File: core/dbt/events/functions.py (hunk 2 above)
Comment (leahwicz) [reply core/dbt/events/functions.py:262]: I'm ok looking into this after the RC3 since I feel like it will be a rare case

**Review by kwigley — APPROVED** [review]

**Comment by leahwicz** [comment]
@nathaniel-may we can merge w/o the 3.9 Mac test succeeding. It looks like a problem with Actions. The Macs are known to have issues...

**Review by nathaniel-may — APPROVED** [review]

**Pull request merged**